### DISCUSSION TITLE
Way to map drag and drop uploads to a configured media type

### DISCUSSION THREAD

**Discussion by enkelmedia** [umbraco/Umbraco-CMS#13721] [Features and ideas]
We're using some custom media types, most of the time a video-media type.

When someone drags and drops ie a .mov or .mp4-file it gets the "umbracoAutoSelect" document type and are sent to the PostAddFile()-file method on the MediaController.

Inside this method we're comparing the extension against known image types and otherwise we're using the "File" media type as a fallback.

It would be useful to be able to hook in here and setup a mapping between file extensions and media types, something that could be configured during startup, maybe using a collection builder or something.

```csharp

  public class MyComposer : IUserComposer
    {
        public void Compose(Composition composition)
        {            
            composition.MediaMappings().Append<VideoFilesMappings>();
        }
    }
```

Is this something that you would consider?

**Comment by nul800sebastiaan** [comment]
Hi @enkelmedia, I will admit, I read this 3 times and I'm not understanding what you're trying to achieve. 🙈 What is your use case, what is currently impossible that this change would  allow you/others to do?

**Comment by enkelmedia** [comment]
hehe sorry! I'll try to make it more clear =D

There is a feature where one can just drag and drop files to upload them to the media-section, implemented in the "umb-file-dropzone"-directive.

![umb drag and drop](https://user-images.githubusercontent.com/1782524/102716274-202ca580-42db-11eb-87a7-6a75595ebec0.PNG)

This feature posts the file to /umbraco/backoffice/UmbracoApi/Media/PostAddFile

The request body looks something like this
```
currentFolder: -1
contentTypeAlias: umbracoAutoSelect
propertyAlias: umbracoFile
file: (binary)
```

Inside the MediaControllers PostAddFile-method: https://github.com/umbraco/Umbraco-CMS/blob/34e80d86e8c0b754f6b7a02e307f53cb32806bbe/src/Umbraco.Web/Editors/MediaController.cs#L617

The contentTypeAlias "umbracoAutoSelect" means that Umbraco will use "File" as the default media type alias for the uploaded file. If the extension is any of the known image-extensions Umbraco will use "Image" as media type. Otherwise it will use the media type "File" (all which are hardcoded).

https://github.com/umbraco/Umbraco-CMS/blob/34e80d86e8c0b754f6b7a02e307f53cb32806bbe/src/Umbraco.Web/Editors/MediaController.cs#L707

In our case we have created a custom media type for videos, what I'm proposing is a way to map ie a ".mp4"-file to this custom type when it's uploaded.

Is in the MediaController around L707 I'm thinking that we could add a collection of "mappings" between file extensions and media types.

So that the code would look both in the known image types and in this collection of configured mappings, so when a editor drags in a .mp4 file it would be created as a "video" and not as a "file". I just proposed a collection builder to be able to append/change the mappings.

So I'm thinking that we could:
* Create a class and interface that represents the map, ie
```csharp
public class VideoMediaMap : IMediaMap {
   public string[] Extensions => new string[]{"mp4","mov"};
   public string ContentTypeAlias => "video";
}
```
* Create a collection builder (like in the example in the first post)
* Use this collection in the MediaController to map a uploaded file to a media type.
* We could also move the current hardcoded logic for images in to a implementation of this IMediaMap-interface.

Hope this makes it more clear =D

**Comment by Matthew-Wise** [comment]
It would also be great if this took into account restrictions on the parent folder. There are cases where you need multiple "pdf" definitions due to different CMS data linked to a file.

**Comment by Zeegaan** [comment]
Sounds like a great feature request, moving this to discussions 👍

**Comment by hfloyd** [comment]
I think being able to map extensions to MediaTypes for the drag-n-drop is a GREAT idea. +💯

It would also give us an easy way to customize the types which are considered "Image" (rather than relying on the hard-coded Configuration)
https://github.com/umbraco/Umbraco-CMS/blob/5457521f19a367c9b5194bffc020a188ac440bfb/src/Umbraco.Core/Configuration/Models/ContentImagingSettings.cs#L31

https://github.com/umbraco/Umbraco-CMS/blob/5457521f19a367c9b5194bffc020a188ac440bfb/src/Umbraco.Core/Configuration/Models/ContentImagingSettings.cs#L13

**Reply by iOvergaard** [reply]
Hi @hfloyd 

Is there a case in particular you would like to cover, that isn't already covered by setting the extension on the upload field?

Set the file extensions:
![image](https://github.com/user-attachments/assets/8564bf35-3db8-4ea3-a92a-38ae500959cb)

Drag and drop:

https://github.com/user-attachments/assets/d5783362-fa6c-441b-9eff-8ea10bf364b4

**Comment by hfloyd** [comment]
Hi @iOvergaard - I actually came upon this discussion while working on an upgrade project for a client. Reviewing their Media section (which has been growing since v7, I admit) I came across some media which were added as Files, whereas they should be Images. I know that new file formats sometimes arise, and I think the few "webp" images in the Media section were likely added prior to that being widely recognized - and thus they were not recognized as Images. 

I was curious to know which extensions would be considered images, (specifically for the drag-n-drop - which I know is a popular way to add files quickly to the Media section) and whether that was configurable for handling new formats as they arise (or perhaps more obscure formats), additionally, it's possible that a developer would want to *encourage* the use of certain formats for on-site images (or *discourage* other formats), and thus it isn't crazy to think this should be somehow configurable, rather than hard-coded in source.

For my purposes, even a simple AppSettings config option would be fine.

As your video demonstrates, the "Upload" property type allows defining the extensions accepted, but this is not the case with the "Image Cropper" type.

**Reply by iOvergaard** [reply]
I must admit I thought this was possible with the option to set [ImageFileTypes](https://apidocs.umbraco.com/v13/csharp/api/Umbraco.Cms.Core.Configuration.Models.ContentImagingSettings.html#Umbraco_Cms_Core_Configuration_Models_ContentImagingSettings_ImageFileTypes) through the app settings. However, upon further inspection, it turns out this option is not used in V13 and is only shallowly referenced in V15.

There is another field defined on the `IImageUrlGenerator` called **SupportedImageFileTypes** that can be overridden. This field holds the value of the currently accepted extensions from the ImageSharp library that are used to resize images. The field is directly mapped into the V13 Backoffice and used for `acceptedFileExtensions` on the image cropper:

```js
if (uploadProperty.editor === "Umbraco.ImageCropper") {
    acceptedFileExtensions = Umbraco.Sys.ServerVariables.umbracoSettings.imageFileTypes;
} else if (uploadProperty.editor === "Umbraco.UploadField") {
    acceptedFileExtensions = (uploadProperty.config.fileExtensions && uploadProperty.config.fileExtensions.length > 0) ? uploadProperty.config.fileExtensions.map(x => x.value) : null;
}
```

It is a bit cumbersome, but in V13 you can extend the class [ImageSharpImageUrlGenerator](https://apidocs.umbraco.com/v13/csharp/api/Umbraco.Cms.Imaging.ImageSharp.Media.ImageSharpImageUrlGenerator.html) and override the [SupportedImageFileTypes](https://apidocs.umbraco.com/v13/csharp/api/Umbraco.Cms.Imaging.ImageSharp.Media.ImageSharpImageUrlGenerator.html#Umbraco_Cms_Imaging_ImageSharp_Media_ImageSharpImageUrlGenerator_SupportedImageFileTypes) field to add image extensions of your choosing.

I will be digging into this some more to figure out what the way forward is for the now-unused app setting (which I guess was used in the past), and for the V15 Backoffice, which currently does not take this into account.
